File: src/halleffect_controller/src/joystick_republisher.cpp
#include <rclcpp/rclcpp.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp>
#include <std_msgs/msg/float64.hpp>

class TwistStampedRepublisher : public rclcpp::Node
{
public:
    TwistStampedRepublisher(): Node("twist_stamped_republisher")
    {
        this->declare_parameter<std::string>("input_topic", "/cmd_vel");
        this->declare_parameter<std::string>("output_topic", "/halleffect_controller/cmd_vel");

        std::string input_topic = this->get_parameter("input_topic").as_string();
        std::string output_topic = this->get_parameter("output_topic").as_string();

        publisher_ = this->create_publisher<geometry_msgs::msg::TwistStamped>(output_topic, 10);
        rpm_publisher_ = this->create_publisher<std_msgs::msg::Float64>("/motor_control", 10);

        subscription_ = this->create_subscription<geometry_msgs::msg::TwistStamped>(
            input_topic, 10,
            [this](geometry_msgs::msg::TwistStamped::UniquePtr msg) {
                publisher_->publish(*msg);

                // Convert linear velocity to RPM
                double diameter = 0.21; // meters
                double linear_velocity = msg->twist.linear.x; // m/s
                double rpm = (linear_velocity / (M_PI * diameter)) * 60.0;

                std_msgs::msg::Float64 rpm_msg;
                rpm_msg.data = rpm;
                rpm_publisher_->publish(rpm_msg);
            });
    }

private:
    rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr publisher_;
    rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr rpm_publisher_;
    rclcpp::Subscription<geometry_msgs::msg::TwistStamped>::SharedPtr subscription_;
};
int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TwistStampedRepublisher>());
    rclcpp::shutdown();
    return 0;
}

File: install/halleffect_controller/share/halleffect_controller/launch/joystick_republisher.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import TwistStamped

class Republisher(Node):
    def __init__(self):
        super().__init__('twist_stamped_republisher')
        self.sub = self.create_subscription(
            TwistStamped,
            '/cmd_vel/geometryTwistStamped',
            self.listener_callback,
            10
        )
        self.pub = self.create_publisher(
            TwistStamped,
            '/halleffect_controller/cmd_vel/geometryTwistStamped',
            10
        )

    def listener_callback(self, msg):
        self.pub.publish(msg)
        self.get_logger().info('Republished TwistStamped message')

def main(args=None):
    rclpy.init(args=args)
    node = Republisher()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()